### PR TITLE
RFC: Device interface (aka threaded i2c/spi)

### PR DESCRIPTION
**Pull request by lucasdemarchi**
Hi,

This is the first part of the "threaded I2C" work. After some work on it it expanded to be "threaded I2C/SPI" and after considering retaining compatibility with current drivers it expanded to "create a better HAL interface for drivers to use".  In the end this better interface came first and the threaded part can be accomplished with some small changes.

Some reasons to change the interface:
 - We need to change some drivers to a threaded model and keep others without converting. This way we can test each of them separately and ensure they work rather than having an "all or nothing" conversion. For that we must share some data between the old and new implementation, mainly the semaphore for synchronization and the file descriptor to write to the bus.
 - The current way of supporting both I2C and SPI sensors are too laborious. Each driver needs a separate "SerialBus" with functions to write to each of them, take the semaphore, etc. By having an underlying common API we can ditch a lot of code. The trend in the current conversions for drivers supporting both buses is that the driver is around 50%~60% the size of the old driver

In order to support the common API there was a major change to how we write to SPI bus: instead of passing just one transfer struct with both tx and rx with rx shifted by tx's length we are now using 2 transfer structs to the same ioctl like the I2C API. For Linux this will actually have the same effect in the kernel.

Support to allow instantiating the drivers and allow them to fail was also improved removing the hardcodednumber of buses. Now each sensor can be in different buses and it requests its own bus to be opened. By using the I2CManager and SPIManager the common struct for this bus (like semaphore and fd) is shared for all sensors that request the same bus.  You can see that now all the information to use a bus and sensor is coming from `AP_HAL_Boards.h`. This is on purpose, paving the way to a transition to runtime configuration file.

The changes to `AP_HAL` add abstract interfaces and all the implementation is currently done in `AP_HAL_Linux`. This is to allow the drivers to be converted more easily. From what I'm seeing I think we can later share with PX4 the registering and book keeping for buses.

Lastly **this has only been bench tested on minlure**, although the other boards were changed as well in order to keep them compiling. A more careful review of the bus numbers and tests will be needed. This PR is to allow others to give a look and early review.

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
I forgot to mention: the drivers currently converted are: BMP085 (that currently nobody uses), MS5611 and MPU6000. The latter is the more complex one since it supports both SPI and I2C, utilize the `transaction()` method from spi directly and has support for auxiliary bus (I2C over SPI). They were chosen since these are the sensors present in minlure and because of "if the new interface satisfies MPU6000 needs, it's good for any other sensor as well".

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 1]
```diff
@@ -210,14 +213,14 @@
...
 #define HAL_BOARD_TERRAIN_DIRECTORY "/data/ftp/internal_000/APM/terrain"
 #define HAL_INS_DEFAULT HAL_INS_MPU60XX_I2C
-#define HAL_INS_MPU60XX_I2C_BUS 2
-#define HAL_INS_MPU60XX_I2C_ADDR 0x68
+#define HAL_INS_MPU60x0_I2C_BUS 2
+#define HAL_INS_MPU60x0_I2C_ADDR 0x68
 #define HAL_COMPASS_DEFAULT HAL_COMPASS_AK8963_I2C
 #define HAL_INS_AK8963_I2C_BUS 1
 #define HAL_COMPASS_AK8963_I2C_ADDR 0x0d
+#define HAL_BARO_DEFAULT HAL_BARO_MS5607
```
Comment (OXINARF) [inline libraries/AP_HAL/AP_HAL_Boards.h]: This should now be HAL_BARO_MS5607_I2C right?

**Comment by OXINARF** [comment]
I've gone through all changes and the only obvious bug I've seen was what I pointed out. I'm far from an expert, especially in the driver department, so it's probable I've missed something.

**Comment by jberaud** [comment]
@OXINARF :+1: 
Tested on bebop quickly, no problem (having fixed the HAL_BARO_MS5607_I2C).
About the overall will to make the whole I2C/SPI device interface more generic with a hierarchy, thanks very much @lucasdemarchi , this is something that's unifying lots of things and something that really needed  to be done.
I have also quickly taken a look and everything looks fine. I will review more in details in the upcoming days.

File: libraries/AP_HAL/AP_HAL_Boards.h (hunk 1 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/AP_HAL_Boards.h]: yep! thanks very much

**Comment by lucasdemarchi** [comment]
Replaced by #3544

**Comment by zehortigoza** [comment]
Sorry I should review the new version.

**Pull request closed without merging**